torchgen/gen_functionalization_type.py

```diff
@@ -540,6 +540,11 @@ def emit_inplace_functionalization_body(
         for a in f.func.arguments.flat_all
         if a.type.is_tensor_like() and a.annotation is None
     ]
+    non_mutated_tensor_names = [
+        a.name
+        for a in f.func.arguments.flat_all
+        if a.type == BaseType(BaseTy.Tensor) and a.annotation is None
```

Comment:
What about optional tensors or lists of tensors?

Reply:
so, I was less worried about optional tensors / lists because:
* most of our ops don't work across multiple devices. e.g. `cpu_tensor.copy_(xla_tensor)` is okay, but `cpu_tensor.add_(xla_tensor)` is not.
* there are very few ops that take optional tensors / lists that actually get mutated